cpp/src/parquet/arrow/reader.cc

```diff
@@ -453,229 +483,226 @@ class NestedListReader : public ColumnReaderImpl {
-    found_nullable_child = true;
-    break;
-  }
+  // This method should only be called when this struct or one of its parents
+  // are optional/repeated or it has a repeated child.
+  // Meaning all children must have rep/def levels associated
+  // with them.
+  RETURN_NOT_OK(def_rep_level_child_->GetDefLevels(data, length));
 
-  if (!found_nullable_child) {
+  if (*data == nullptr) {
+    // Only happens if there are actually 0 rows available.
     *data = nullptr;
     *length = 0;
-    return Status::OK();
   }
+  return Status::OK();
+}
 
-  // Look at the rest of the children
-
-  // When a struct is defined, all of its children def levels are at least at
-  // nesting level, and def level equals nesting level.
-  // When a struct is not defined, all of its children def levels are less than
-  // the nesting level, and the def level equals max(children def levels)
-  // All other possibilities are malformed definition data.
-  for (; child_index < static_cast<int>(children_.size()); ++child_index) {
-    // Child is non-nullable, and therefore has no definition levels
-    if (!children_[child_index]->field()->nullable()) {
-      continue;
-    }
+Status StructReader::GetRepLevels(const int16_t** data, int64_t* length) {
+  *data = nullptr;
+  if (children_.size() == 0) {
+    *length = 0;
+    return Status::Invalid("StructReader had no childre");
+  }
 
-    auto& child = children_[child_index];
-    int64_t current_child_length;
-    RETURN_NOT_OK(child->GetDefLevels(&child_def_levels, &current_child_length));
-
-    if (child_length != current_child_length) {
-      std::stringstream ss;
-      ss << "Parquet struct decoding error. Expected to decode " << child_length
-         << " definition levels"
-         << " from child field \"" << child->field()->ToString() << "\" in parent \""
-         << this->field()->ToString() << "\" but was only able to decode "
-         << current_child_length;
-      return Status::IOError(ss.str());
-    }
+  // This method should only be called when this struct or one of its parents
+  // are optional/repeated or it has repeated child.
+  // Meaning all children must have rep/def levels associated
+  // with them.
+  RETURN_NOT_OK(def_rep_level_child_->GetRepLevels(data, length));
 
-    DCHECK_EQ(child_length, current_child_length);
-    for (int64_t i = 0; i < child_length; i++) {
-      // Check that value is either uninitialized, or current
-      // and previous children def levels agree on the struct level
-      DCHECK((result_levels[i] == -1) || ((result_levels[i] >= struct_def_level_) ==
-                                          (child_def_levels[i] >= struct_def_level_)));
-      result_levels[i] =
-          std::max(result_levels[i], std::min(child_def_levels[i], struct_def_level_));
-    }
+  if (data == nullptr) {
+    // Only happens if there are actually 0 rows available.
+    *data = nullptr;
```

Comment:
You're dereferencing a null pointer (see `if` condition above).

Reply:
yeah, removed.  we shouldn't need this.